src/array.rs

```diff
@@ -136,13 +136,23 @@ impl<'a, T: Element, D: Dimension> FromPyObject<'a> for &'a PyArray<T, D> {
             }
             &*(ob as *const PyAny as *const PyArray<T, D>)
         };
-        let dtype = array.dtype();
-        let dim = array.shape().len();
-        if T::is_same_type(dtype) && D::NDIM.map(|n| n == dim).unwrap_or(true) {
-            Ok(array)
-        } else {
-            Err(ShapeError::new(dtype, dim, T::DATA_TYPE, D::NDIM).into())
+
+        let src_dtype = array.dtype();
+        let dst_dtype = T::get_dtype(ob.py());
+        if !src_dtype.is_equiv_to(dst_dtype) {
+            // FIXME: this is unsound - we're only checking that the dtype is 'O'
```

Comment:
I don't think this is really correct. This is only applies to object types and we are not providing any `Element` impls besides `Py<PyAny>` which is safe. So together with #255, I don't see any soundness issues here?

Reply:
You're implying that it's sound because the `Element` trait is unsafe to impl? It doesn't make it not unsound - we're still technically providing a footgun, no?

Reply:
> You're implying that it's sound because the Element trait is unsafe to impl?

That is the technical argument, yes.

> It doesn't make it not unsound - we're still technically providing a footgun, no?

It does make it not sound if that is part of the traits contract. In #255, it is expressed that one cannot implement an object-like `Element` for anything but `PyObject` for now as we do not check each elements dynamic type. We might relax this eventually, but as it stands, there is no unsoundness after #255 is applied.

Reply:
Fair enough, I'll remove the note.

It would probably be nice to have other-than-PyAny support at some point, but it's good for now.